src/index.js

```diff
@@ -21,20 +21,59 @@ function crossEnv(args, options = {}) {
         env,
       },
     )
-    process.on('SIGTERM', () => proc.kill('SIGTERM'))
-    process.on('SIGINT', () => proc.kill('SIGINT'))
-    process.on('SIGBREAK', () => proc.kill('SIGBREAK'))
-    process.on('SIGHUP', () => proc.kill('SIGHUP'))
-    proc.on('exit', (code, signal) => {
-      let crossEnvExitCode = code
-      // exit code could be null when OS kills the process(out of memory, etc) or due to node handling it
-      // but if the signal is SIGINT the user exited the process so we want exit code 0
-      if (crossEnvExitCode === null) {
-        crossEnvExitCode = signal === 'SIGINT' ? 0 : 1
+
+    // See https://github.com/jtlapp/node-cleanup
+    //
+    // > When you hit Ctrl-C, you send a SIGINT signal to each process in the
+    // > current process group. A process group is set of processes that are
+    // > all supposed to end together as a group instead of persisting
+    // > independently. However, some programs, such as Emacs, intercept and
+    // > repurpose SIGINT so that it does not end the process. In such cases,
+    // > SIGINT should not end any processes of the group.
+    //
+    // Delegate decision to terminate to child process, if the child exits on
+    // SIGINT then the `child.on('exit')` callback will be invoked, re-raising
+    // the signal to the parent process
+    const delegateSignalToChild = signal => () => {
+      // SIGINT is sent to all processes in group, no need to delegate.
+      if (child && signal !== 'SIGINT') {
+        process.kill(child.pid, signal)
+      }
+    }
+
+    const sigtermHandler = delegateSignalToChild('SIGTERM')
+    const sigintHandler = delegateSignalToChild('SIGINT')
+    const sigbreakHandler = delegateSignalToChild('SIGBREAK')
+    const sighupHandler = delegateSignalToChild('SIGHUP')
+    const sigquitHandler = delegateSignalToChild('SIGQUIT')
+
+    process.on('SIGTERM', sigtermHandler)
+    process.on('SIGINT', sigintHandler)
+    process.on('SIGBREAK', sigbreakHandler)
+    process.on('SIGHUP', sighupHandler)
+    process.on('SIGQUIT', sigquitHandler)
+
+    child.on('exit', (exitCode, signal) => {
+      // Child has decided to exit.
+      child = null
+      process.removeListener('SIGTERM', sigtermHandler)
+      process.removeListener('SIGINT', sigintHandler)
+      process.removeListener('SIGBREAK', sigbreakHandler)
+      process.removeListener('SIGHUP', sighupHandler)
+      process.removeListener('SIGQUIT', sigquitHandler)
+
+      if (exitCode !== null) {
+        // Calling process.exit is not necessary most of the time,
+        // see https://nodejs.org/api/process.html#process_process_exit_code
+        process.exitCode = exitCode
+      }
+      if (signal !== null) {
+        // Pass through child's signal to parent.
+        // SIGINT should not be transformed into a 0 exit code
+        process.kill(process.pid, signal)
```

Comment:
Do I understand correctly that the code does the following?
- when the parent process (`cross-env`) receives a `SIGTERM` signal, it will handle it by sending it to the child process
- then when the child process exits and its exit status says it's been killed with a signal, the parent sends that signal to itself?
- that means that the parent process receives the signal two times: first from external source, handled by a custom handler that forwards it to the child, second time from itself, and handled by the default handler, as the custom handler was removed a moment ago?
- the second `SIGTERM`, with custom handler removed, will cause the parent process to terminate, as that's what the default handler does?
- the parent process will end with an exit status that says it was terminated with `SIGTERM`?

The `SIGINT` handling seems to be a bit special: the parent process' custom handler simply ignores it, and relies on the fact that the child process received the `SIGINT` signal, too. Then in the `child.on('exit')` callback, it will uninstall the "ignore" handler and will send `SIGINT` to itself. Then it will finally terminate with `SIGINT` signal exit status.

One case that looks a bit suspicious: if I kill the parent process explicitly, with `kill -s SIGINT <parent-pid>`, then it will ignore the signal and do nothing. Both processes continue to run.

Reply:
Yes your understanding is correct for the example using `SIGTERM` and all other non-SIGINT signals that are listened for.

For the example of SIGINT, all I can say is the documentation I found indicates this is what happens on Unix systems. I haven't tested this on that platform. If you can find a definitive guide to signal handling I"m happy to adjust the code.

On Windows there is no such thing as signals, they are emulated by Node [Signal Events](https://nodejs.org/api/process.html#process_signal_events)

Did you try the `kill -s SIGINT <parent-pid>`? As I'm on windows I dont have those utilities.